docs/deployment-ux.md

```diff
@@ -0,0 +1,410 @@
+every instance in the target Instance Group:
+
+``` {.shell}
+$ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+SomeEnvironment/Prod:1    Undeploying  0         5         5            5        
+SomeEnvironment/Prod:2    Deploying    5         0         0            0        
+```
+
+Once the deployment reaches a stead state, the old revision is marked
+Inactive, and the new revision is marked Active:
+
+``` {.shell}
+$ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+SomeEnvironment/Prod:1    Inactive     0         0         0            0        
+SomeEnvironment/Prod:2    Active       5         5         5            5        
+```
+
+### Updating the Instance Group
+
+If you don't specify all attributes, you must specify the revision you
+wish to base your changes on. This prevents potential conflicts if
+multiple callers attempt to update an environment at the same time.
+
+``` {.shell}
+$ aws ecs update-environment --environment SomeEnvironment/Prod --revision 1 --instance-group-query "attribute:stack in (prod, gamma)"
+{ "EnvironmentRevision": 2 }
+```
+
+At this point, nothing has actually happened to your cluster. You can
+see this by showing your environment's status:
+
+``` {.shell}
+$ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+REVISION                  STATUS     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+SomeEnvironment/Prod:1    Active     5         5         5            5        
+SomeEnvironment/Prod:2    Created    0         0         0            0        
+```
+
+In order to make this configuration Active, you first have to Deploy the
+revision:
+
+``` {.shell}
+$ aws ecs deploy-environment --environment-name SomeEnvironment/Prod --environment-revision 2
+{ "DeploymentId": "SomeEnvironment/Prod:2@2017090611472301" }
+```
+
+This will mark the revision as Deploying, and start launching tasks to
+meet the deployment configuration's constraints. In the case of a Daemon
+environment, the desired target is to have one copy of the task for
+every instance in the target Instance Group. Since the size of the
+instance group is now different because it was changed, that's reflected
+in the DESIRED column for the new revision:
+
+``` {.shell}
+$ aws ecs describe-environment-status --environment-name SomeEnvironment/Prod --table
+REVISION                  STATUS       DESIRED   CURRENT   UP-TO-DATE   AVAILABLE
+SomeEnvironment/Prod:1    Undeploying  0         5         5            5        
+SomeEnvironment/Prod:2    Deploying    6         0         0            0        
+```
+
+Note that even though this doesn't change the task definition, it will
+still stop and restart tasks. (TODO: Can we make it smarter, so that it
```

Comment:
This behaviour (stopping/starting all tasks for an instance group change) is not what we think customers will want (*"I just added an extra instance to my instance group, why is it killing and restarting every single task??"*). It would make much more sense to customers if we seamlessly detected that there was no other change, and just roll out new tasks to new instances that are newly part of the group, or terminate tasks on instances that are no longer in the group.

Our limitation to making this smarter is that we can't attach mutable metadata to instances to identify which environment revision it belongs to. We think we can do something smart to implement it this way anyway, so we'll aim for the better, more efficient solution.